### DISCUSSION TITLE
用GeeksforGeeks學習資料結構和演算法

### DISCUSSION THREAD

**Discussion by KH1126** [1101-GS4719-C/Python-Community-GS4719-C#1911] [General]
上學期選修了資料結構課程，在教到Tree章節時，對紅黑樹和AVL Tree有一些概念上的問題，諸如紅黑樹是怎麼操作，他的規則為何?
因此通過在網上搜尋查到GeeksforGeeks這個網站。
簡單來說，你可以在這個網站上學習到很多關於計算機知識的內容，像是資料結構、演算法或者是資料庫管理的知識。
以下為網站連結:[(**https://www.geeksforgeeks.org/**)](https://www.geeksforgeeks.org/)
這裡來用GeeksforGeeks上教到linked list這個資料結構來說明。

![螢幕擷取畫面 2024-03-06 122901](https://github.com/NCU-GS4719-Python/Python-Community/assets/100942642/2c7edd58-fcf8-46e3-bb6f-3af2a2dde4b4)
他會先介紹linked list所會用到的概念、組成，例如:鍊錶是一種線性資料結構，其中資料不儲存在連續位置(記憶體位置)，而是使用指標連結(串接)，形成一系列相連的節點，每個節點儲存資料和下一個節點的位址。

在介紹完linked list概念後，也會有一些運用到linked list的問題來輔助學習。
案例如下:

![螢幕擷取畫面 2024-03-06 123616](https://github.com/NCU-GS4719-Python/Python-Community/assets/100942642/6049d418-d59d-489c-9b9a-f53a739e06b5)

這個網站會提供問題所需要用到的概念為何，並附上問題實作解法可供參考。

![螢幕擷取畫面 2024-03-06 123738](https://github.com/NCU-GS4719-Python/Python-Community/assets/100942642/364898f0-bee4-4018-ba74-e54e01e75ab2)

我認為GeeksforGeeks是一個非常有價值的網站，無論是初學者或是專業人士，都可以在這個網站上找到對他們有用的知識和資訊。
希望這個資訊對大家有幫助。

**Comment by jim1743** [comment]
感謝分享，我覺得這個網站對於我之後學習資料結構相當有幫助。

**Comment by William08290829** [comment]
感謝分享 之前在資料結構這邊的確比較難以理解 太電了老哥

**Comment by EPHMR0524** [comment]
我覺得這些資訊相當有用，這學期個人越級打怪沒修資結就去修演算法，蠻需要些好方法去快速學習資結的內容
利用geekforgeeks還能做到在學習完觀念之後馬上有一個例子可以學習，對資料結構的概念能理解得更快，我相信再使用了這個網站後我可以更快地更上進度。

**Comment by YifenLin** [comment]
目前還沒修過資料結構 但有想要學習相關的知識 這個網站是一個很好學習的管道 感謝分享～

**Comment by benbebobo149** [comment]
資料結構的確是一門非常重要的課程，因為沒有運用好資料結構，我可能會陷入繁瑣的重複運算。我也強烈推薦更多人對紅黑樹進行深入研究，因為許多資料庫的底層都是用這個算法來實作。對於這位同學的分享，我感激不盡。 GeeksforGeeks提供了豐富的資源，我鼓勵更多人利用這個寶貴的學習平台來加強對資料結構等知識的理解。

**Comment by lyykey** [comment]
謝謝版主分享這個資訊，準備要開始學習資料結構，剛好有地方能夠入門了。

**Comment by happy246810** [comment]
我雖然尚未接觸過資料結構相關的課程，但我認為這個網站能讓我更快速且更輕鬆地了解其概念。並且作為往後學習計算機科學的預先準備，這對我非常有幫助。非常感謝版主分享這個寶貴的資訊。

**Comment by yyujimmy** [comment]
以下是我規則理解整理後的看法:
紅黑樹是一種自平衡二元搜尋樹，主要用於實現資料的插入、刪除和查找等操作，確保樹的高度維持在一個較低的水平，以確保操作的效率。
紅黑樹的規則如下：
1. 每個節點要麼是紅色，要麼是黑色。
2. 根節點是黑色。
3. 每個葉節點（NIL 節點）是黑色的。
4. 如果一個節點是紅色的，則其子節點必須是黑色的（紅黑交替）。
5. 從任一節點到其每個葉節點的路徑都包含相同數量的黑色節點。

紅黑樹透過對節點的顏色進行旋轉和變換，來維護這些規則，保持樹的平衡。紅黑樹的操作包括插入、刪除和查找等，這些操作都會經過特定的旋轉和調整來確保樹的平衡性。透過遵守紅黑樹的規則，可以確保樹的高度保持在一個較低的水平，提高樹的查找效率。

而你提供的這個網站，我瀏覽過後，覺得非常有用，也從中得到許多課外知識，再稍微補充一下這個網站的資訊，GeeksforGeeks 是一個電腦科學教育網站，提供廣泛的資訊和教學內容，涵蓋了不同領域的程式設計主題。網站提供了豐富的程式設計問題和解答、技術文章、教學影片等資源，幫助學生和專業人士提升他們的程式設計技能和知識。

以上，希望能幫助大家。

**Reply by KH1126** [reply]
感謝分享。

AVL樹和紅黑樹在不同的場景下各有其微弱優勢，但幾乎不可能出現一種在所有方面都壓倒另一種的情況。增刪查改是一個充滿了權衡考量的過程。

在這裡，我再分享一篇介紹AVL樹和紅黑樹的比較文章。
[https://zhuanlan.zhihu.com/p/93369069](https://zhuanlan.zhihu.com/p/93369069)

**Comment by sunset513** [comment]
我認為資料結構是一門十分重要的課程，這是一個承先啟後的關鍵課程。在大一學過了基礎程式設計後，我學會了一些基本語法，並懂得簡單地使用。在大二學習了資料結構後，我對於程式語言的瞭解又更加地深入了，了解了許多設計的原理。在學習了這門課程後，對於之後艱難的計算機知識，我能夠更好的去做吸收。希望我們都能持續進步，在學習計算機科學的路上不斷前進。

**Comment by OG-Matcha** [comment]
資料結構作為資工六科之一確實是接觸程式的必經之路，我個人認為資料結構不只是用在應付課程，更是許多演算法的基礎，以及很多延伸應用的基石，他會變成你在寫程式時的一個良好觀念，會影響到你撰寫程式時的結構及效率。

**Comment by wanderer657** [comment]
做為修習演算法之前的基石，資料結構相當重要，感謝分享其相關知識，這篇所提到的link listed與個人過往接觸到的內容有所相似，相信不論對於他人或我，都會重新啟發、激勵對於資料結構的探索。

**Comment by SusuTima** [comment]
上學期在修資結的時候，也很常用「iT邦幫忙」和「Medium」這兩個網站，上面有很多資結和演算法的教學，幫我建立了許多程式概念的基礎。

**Comment by SusuTima** [comment]
資結、演算法也可以去聽台大林軒田教授的「Data Structures and Algorithms」，老師在裡面講解的很細膩，對初學的我給予了很多幫助。

**Comment by poiu6fi** [comment]
感謝分享，之後有考慮修資工的資結，怕會跟不上他們的進度，有了這個應該可以更好理解

**Comment by HirokiMantissa** [comment]
對資結的印象都是在畫畫，可能是那堂課課前考蠻多畫畫的題目，但畫得出代表對架構有了解，畢竟亂畫很容易發現。

**Comment by hbh10601** [comment]
剛好學期末要考演算法，原本一直都在Chat，看來有個更專業的地方可以補充加強學習相關知識不足的地方ㄌ，感謝同學推薦！
